Julietshen/update roadmap (#52)
roostorg/community: roadmap.md

# ROOST Project Roadmap

This roadmap outlines the short-term development priorities for ROOST's open source trust and safety infrastructure and specifically addresses the needs the ROOST team is currently equipped to be building against. We look forward to ideas, suggestions, and feedback from the community; any feature requests not on the roadmap will be logged in a separate [“wishlist”](https://github.com/orgs/roostorg/discussions/50) that collects the systems and tools most wanted across the open source safety ecosystem. Whether you're evaluating ROOST for your platform, looking to contribute, or coordinating work across project teams, this document will help you understand where we're headed and how to get involved.

> [!NOTE]

> > Timelines are based on assumptions on team sizing and overall engineering contributions. As of April 2026, we have received numerous community contributions but have not yet hired full-time engineers for the ROOST development team. [Are you interested? Apply here!](https://roost.tools/careers)

Reviewer: ```suggestion
> Timelines are based on assumptions on team sizing and overall engineering contributions. As of May 2026, we have received numerous community contributions but have not yet hired full-time engineers for the ROOST development team. [Are you interested? Apply here!](https://roost.tools/careers)
```

| :---------------------------------------------------------- | :---------------------------------------------------------------------------- |
| Built and donated by [Discord](https://discord.com/blog/osprey-open-sourcing-our-rule-engine) and open sourced through ROOST | ROOST-acquired IP from [Cove](https://getcove.com/)                           |
| Human-crafted rules actioned at scale                       | Flexible review tool for labeling multiple formats (ie. content and accounts) |
| High QPS processing for streaming and batched data          | Queue orchestration, audit trails, reviewer wellness features                 |
| Open-ended investigation                                    | Configurable actions, entities, and dashboards                                |
| UI for analysts to identify abuse patterns and signals      | Automated routing of tasks into queues                                        |
| Sync and async rule creation and execution                  | Complete CSAM detection and reporting system                                  |

## Preparing for AI-Powered Safety

> As AI increasingly gets used by bad actors, it's critical that those who work on online safety have access to the same degree of technology to protect online communities. In addition, product improvements are shipping rapidly and adversarial behavior is evolving with them. AI can make safety capabilities accessible to teams that could never have built them from scratch. [Our blog post on deploying AI agents for safety](https://roost.tools/blog/how-to-deploy-ai-agents-for-safety/) covers the motivation and patterns in more depth, including production examples from Block and Notion. In order to prepare our projects for AI-powered features, there are two foundational components needed:

Reviewer: I have this persistent nagging issue with the breath and imprecision of "AI" as a term, roping in several different technologies. Maybe that's a lost battle, but I wanted to flag it.

I also worry that this could read at a glance too much like, "we're going all-in on AI!" versus "we're ensuring our tools work well with LLMs/agentic integrations should platforms choose to use them."

If I were a platform or organization _opposed_ to the use of LLMs and/or generative AI (for example, several platforms designed for open source communities, artist/maker communities, writers, etc.), this could be a turn-off to me. So I want to make sure we're balancing things a bit:

- We know that generative AI use is increasingly used by bad actors
- Online safety teams need access to the best tools, including "AI" technologies like LLMs and agentic workflows if they choose
- If platforms/orgs do choose to use ROOST projects with AI technologies, these are things we know we need to do to improve our projects (and online safety tools in general) to enable this

Author: Thanks! I think I can also add some language introducing why we're adding this update, as a way to provide more detail to the previously very ambiguous "add AI features to our tools" that was previously in our roadmap. 

I think using AI for safety is something we should not shy away from; we already have a disclaimer in the Roadmap that says that those who don't want to use AI can use earlier versions that are tagged as pre-AI (which we should plan/discuss, since we're getting closer to this on Osprey).

### Data Abstraction Layer

> Today, both Coop and Osprey only see data pushed through an input event stream. Useful context for investigations and review decisions (account history, reputation scores, relationships between entities) often lives in other systems within each org, and every org has a different data model. The Data Abstraction Layer gives both tools a shared way to understand org-specific data models and pull in richer context. We're proposing three primitives: an Entity Graph Declaration, a Content Enrichment interface, and a Graph Query Protocol. See the [detailed proposal in the roadmap discussion topic](https://github.com/orgs/roostorg/discussions/44) for more.

Reviewer: I suggest making this a bit more clearly forward-looking versus sounding like something that already exists.

```suggestion
Today, both Coop and Osprey only see data pushed through an input event stream. Useful context for investigations and review decisions (account history, reputation scores, relationships between entities) often lives in other systems within each org, and every org has a different data model. A common _data abstraction layer_ would give both tools a shared way to understand org-specific data models and pull in richer context. We're proposing three primitives: an Entity Graph Declaration, a Content Enrichment interface, and a Graph Query Protocol. See the [detailed proposal in the roadmap discussion topic](https://github.com/orgs/roostorg/discussions/44) for more.
```

### Safety Decision Taxonomy

> Policy enforcement for individual content is well understood, but policies for complex patterns (sequences of behaviors, correlated accounts, coordinated content) are rarely formally defined. These are the patterns we expect from AI-assisted adversarial behavior, and they continuously mutate. The Safety Decision Taxonomy is a shared language between humans and AI safety agents that defines how to handle known violations, borderline cases, and new behaviors that need classification. It includes a self-extending feedback loop so the system adapts as threats evolve. See the [detailed proposal in the roadmap discussion topic](https://github.com/orgs/roostorg/discussions/44) for more.

Reviewer: A few suggestions focused on language around "AI" and (as above) adjusting this to be a bit more clearly forward-looking.

```suggestion
Policy enforcement for individual content is well understood, but policies for complex patterns (sequences of behaviors, correlated accounts, coordinated content) are rarely formally defined. These are the patterns we expect from AI-assisted adversarial behavior, and they continuously mutate. A _safety decision taxonomy_ would be a shared language between humans and automated tools including agentic workflows that defines how to handle known violations, borderline cases, and new behaviors that need classification. It would include a self-extending feedback loop so the system adapts as threats evolve. See the [detailed proposal in the roadmap discussion topic](https://github.com/orgs/roostorg/discussions/44) for more.
```

**Core features:**

- [Code-free rules management through UI ](https://github.com/roostorg/osprey/issues/58)
- [Shadow mode for testing rules before production](https://github.com/roostorg/osprey/issues/81)
- [Batch processing for historical analysis](https://github.com/roostorg/osprey/issues/208)
- [Enhanced drill-down capabilities for complex cases](https://github.com/roostorg/osprey/issues/209)

**Infrastructure modernization and security hardening**

Alongside feature work, we are modernizing Osprey's frontend (Antd 5 migration plus incremental UI improvements), backend concurrency (migrating workers from gevent to asyncio), and plugin extensibility (custom sinks, validation exporters, and configurable result stores so Osprey fits into a wider range of org infrastructures), while addressing dependency hygiene to keep the project secure and contributor-friendly. These changes mirror the [simplification work happening in Coop](https://github.com/roostorg/coop/discussions/123) and ensure both projects maintain a consistent, modern development experience.

**Early exploration**

> * Entity Graph Declaration and Content Enrichment integration (see Data Abstraction Layer)

Reviewer: ```suggestion
* Entity Graph Declaration and Content Enrichment integration (see [Data Abstraction Layer](#data-abstraction-layer) above)
```

* Pattern detection for emerging threats
* An ML Platform can be used to convert Osprey rules into continuous learning classifiers

These features were prioritized after shadowing analysts at Discord and Bluesky to understand their actual workflows of version 1.0. In 1.0, rule updates require pull requests and GitHub file management, which creates friction for analysts who aren’t used to making code changes and slows rule creation/deployment. Shadow mode helps ensure rules are catching what they intend to target, and batch processing allows for deeper investigations beyond near-real time. These changes make Osprey work better for current users while expanding who can effectively use it.

### Next - 2027 and Beyond

> **Goal**: Use AI to surface unknown threats and patterns that human analysts might miss. We've heard from Osprey users who already combine Osprey with an MCP to query and create rules using natural language; these features formalize and extend that pattern. The AI landscape is moving fast, and product improvements could outpace the evolution of safety infrastructure, especially for smaller companies. Osprey's AI features are designed to keep pace.

Reviewer: Continuing to try to balance those perspectives I mentioned above, clearly separate out the goal itself from the context, and make a little more concise.

```suggestion
**Goal**: Enable and better support use of AI technologies to surface unknown threats and patterns that human analysts might miss to keep pace with evolving threats.

We've heard from adopters who already combine Osprey with an MCP to query and create rules using natural language; we aim to formalize and extend those patterns, especially to better equip smaller organizations.
```

**Core features:**

> * AI-powered interface for querying data (built on the Data Abstraction Layer)

Reviewer: ```suggestion
* Improved interface for querying data, i.e. using AI technologies like LLMs (built on the [Data Abstraction Layer](#data-abstraction-layer))
```

> * Safety Decision Taxonomy enforcement (pattern classification and feedback loop)

Reviewer: ```suggestion
* [Safety Decision Taxonomy](#safety-decision-taxonomy) enforcement (pattern classification and feedback loop)
```

* Unsupervised learning (clustering, anomaly detection)
* Automated signal combination for improved recall

These features are exploratory pending v1.x feedback and resourcing. More information is needed, like whether production deployments reveal specific investigation gaps worth targeting before general-purpose AI assistance.

## Coop: Review and Enforcement [(source code)][Coop]

![Screenshot of Coop](https://github.com/roostorg/coop/blob/main/docs/images/coop-overview.png?raw=true)

**Current status:** 🟢 v0.1 released, v1 in development

**Project goal:** Provide human-centered review infrastructure that works at scale while protecting reviewer wellbeing, ensuring consistent policy enforcement, and meeting complex reporting obligations.

> **Solution:** Coop is a flexible review console for Trust & Safety across different harm types. Built as an open source refactor of Cove ([a commercial tool whose IP was acquired by ROOST](https://roost.tools/blog/roost-announces-coop-and-osprey-free-open-source-trust-and-safety-infrastructure-for-the-ai-era/)), it provides queue orchestration, context-rich review interfaces, reviewer wellness features, and enforcement workflows. Our V0.1 release includes enhanced specialized child safety workflow functionality alongside the core review capabilities.

Reviewer: Tiniest nit: consistent styling of `v` for version.

```suggestion
**Solution:** Coop is a flexible review console for Trust & Safety across different harm types. Built as an open source refactor of Cove ([a commercial tool whose IP was acquired by ROOST](https://roost.tools/blog/roost-announces-coop-and-osprey-free-open-source-trust-and-safety-infrastructure-for-the-ai-era/)), it provides queue orchestration, context-rich review interfaces, reviewer wellness features, and enforcement workflows. Our v0.1 release includes enhanced specialized child safety workflow functionality alongside the core review capabilities.
```

* Security hardening (XSS/CSRF cookie fixes, signing key rotation)

Organizations need review infrastructure that works for all violation types across accounts and content. Contextual interfaces came from Cove’s commercial learnings and the ROOST team’s own T&S experience about what reviewers actually need to make informed decisions. Reviewer wellness is critical for reducing trauma for T&S workers and belongs in the initial release rather than an afterthought. Community feedback has validated our initial focus on child safety technology.

### v1.0 - 2026

**Goal**: Reduce deployment complexity, harden security, and modernize the developer experience so that Coop is easier to self-host, safer to run, and more accessible to contributors. Establish a cleaner foundation for AI-powered features planned in subsequent releases.

**Core features:**

* [Infrastructure simplification](https://github.com/roostorg/coop/discussions/123) (reduced service count, optional distributed infrastructure, PostgreSQL as lightweight default, ORM migration from Sequelize to Kysely)
* Security hardening and dependency modernization
* Frontend toolchain migration and bundle size reduction

> * Improved local development setup and contributor onboarding

Reviewer: Probably worth explicitly calling out docs as part of this

```suggestion
* Improved local development setup, contributor onboarding, and documentation
```

Community feedback highlighted that Coop's deployment complexity and architectural choices inherited from its commercial origins create friction for self-hosted adoption and contribution. These changes make Coop accessible to smaller teams and individual contributors.

### v1.x features - 2026

**Goal**: Build systematic quality into review workflows, create feedback loops between review decisions and investigation systems, expand child safety tooling for international cooperation, and lay the data foundation for AI-powered features.

**Core features:**

* [In-tool Quality Assurance (QA) for reviewer decisions](https://github.com/roostorg/coop/issues/208)
* [Expanded search](https://github.com/roostorg/coop/issues/209)
* UI improvements
* [Semantic hash detection](https://github.com/roostorg/coop/issues/210)
* [Integrated feedback loops with Osprey](https://github.com/roostorg/coop/issues/211)
* [INHOPE Universal Schema](https://github.com/roostorg/coop/issues/212)

> * Data Abstraction Layer integration (Entity Graph Declaration, Content Enrichment)

Reviewer: ```suggestion
* [Data Abstraction Layer](#data-layer-abstraction) integration (Entity Graph Declaration, Content Enrichment)
```

**Early exploration**

> * Native support for open weight models — scoping UX and feature requirements for selecting and running a model (e.g., from HuggingFace) directly within Coop, going beyond the current API-style ML/AI abstraction

Reviewer: ```suggestion
* Native support for open weight models: scoping UX and feature requirements for selecting and running a model (e.g., from HuggingFace) directly within Coop, going beyond the current API-style ML/AI abstraction
```

QA features emerged from conversations with operations teams who lack systematic approaches beyond spreadsheet-based audits. Improved search enables investigations for ad-hoc escalations and spot-checks. Integrated feedback loops with Osprey create a continuous improvement cycle where review decisions help refine detection rules. INHOPE mapping extends our NCMEC work to international child safety hotlines, recognizing that abuse crosses borders.

### Next - 2027

> **Goal**: Help users make faster, more consistent decisions by providing AI-powered policy reasoning and context summaries, and help teams discover and respond to new harm patterns.

Reviewer: ```suggestion
**Goal**: Help users make faster, more consistent decisions by offering AI-assisted policy reasoning and context summaries, and help teams discover and respond to new harm patterns.
```

**Core features:**

- Agentic review with structured reasoning (human-in-the-loop)

> - Safety Decision Taxonomy integration (routing based on violative, needs review, needs classification dispositions)

Reviewer: ```suggestion
- [Safety Decision Taxonomy](#safety-decision-taxonomy) integration (routing based on violative, needs review, needs classification dispositions)
```

- Review technical requirements and integration patterns
- Join our [Discord server] to keep up with office hours, discussions, and ask questions

> - Join [office hours](https://roostorg.github.io/community/meetings.html) to discuss your specific situation

Reviewer: We run both office hours and working groups. Plus, link to the bare path so it will handle the changes in #54 cleanly.

```suggestion
- Join [public community meetings](https://roostorg.github.io/community/meetings) to discuss your specific situation
```
